components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java

```diff
@@ -311,23 +312,172 @@ public void deleteUser(String userId) throws NotFoundException, CharonException
+
+            return userNames;
+        } catch (UserStoreException e) {
+            throw new CharonException(String.format("Error while listing users from domain: %s.", domainName), e);
+        }
+    }
+
+    private String[] getAllUsernamesByAllDomainNames() throws CharonException {
+
+        String[] userNames = null;
+
+        try {
+            Map<String, String> scimToLocalClaimsMap = SCIMCommonUtils.getSCIMtoLocalMappings();
+            String userIdLocalClaim = scimToLocalClaimsMap.get(SCIMConstants.CommonSchemaConstants.ID_URI);
             if (StringUtils.isNotBlank(userIdLocalClaim)) {
                 userNames = carbonUM.getUserList(userIdLocalClaim, "*", null);
             }
-            if (userNames != null && userNames.length != 0) {
+
+            return userNames;
+        } catch (UserStoreException e) {
+            throw new CharonException("Error while listing users across all domains. ", e);
+        }
+    }
+
+    private String[] getAllUserNamesByAllDomainNames(int offset, int limit, String sortBy, String sortOrder)
+            throws CharonException {
+
+        String[] userNames = null;
+        int initialLimit = limit;
+        int cumulativeCount = 0;
+        UserStoreManager tempCarbonUM = carbonUM;
+
+        while (tempCarbonUM != null) {
+            String[] tempUserNames;
+            int tempCount;
+            // If carbonUM is not an instance of Paginated User Store Manger we can't get the domain name.
+            if (tempCarbonUM instanceof PaginatedUserStoreManager) {
+                String domainName = tempCarbonUM.getRealmConfiguration().getUserStoreProperty("DomainName");
+                tempUserNames = getUsernamesByDomainName(offset, limit, sortBy, sortOrder, domainName);
+
+                if (tempUserNames.length != 0) {
+                    tempCount = tempUserNames.length;
+                    cumulativeCount = cumulativeCount + tempCount;
+                    if (initialLimit > cumulativeCount) {
+                        userNames = (String[]) ArrayUtils.addAll(userNames, tempUserNames);
+                        limit = initialLimit - cumulativeCount;
+                        offset = 0;
+                    } else if (initialLimit == cumulativeCount) {
+                        userNames = (String[]) ArrayUtils.addAll(userNames, tempUserNames);
+                        break;
+                    }
+                } else {
+                    String[] tempUsers = getAllUsernamesByDomainName(domainName);
+                    if (ArrayUtils.isEmpty(tempUsers)) {
+                        offset = offset;
+                    } else {
+                        if (offset >= tempUsers.length) {
+                            offset = offset - tempUsers.length;
+                        }
+                    }
+                }
+
+                tempCarbonUM = tempCarbonUM.getSecondaryUserStoreManager();
```

Comment:
Did you test this logic by adding primary and few secondary user stores then try to iterate user stores? This logic was identified as the bug in [https://wso2.org/jira/browse/IDENTITY-7383](https://wso2.org/jira/browse/IDENTITY-7383). The corresponding source code logic  ( [here](https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/blob/6f5c13fcb55f7e45ad6187f68db0d6b48d158780/components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java#L892)) is similar to this logic.

Reply:
Yes. Test with LDAP as primary user store and H2 as the secondary user store. No issue occurred.